File: src/lib/automation/page-objects/FormA28Page.ts
import type { Page } from 'playwright';
import type { FormA28Data } from '@/lib/validation/formA28Schema';
import { getAllMappings, type FieldMapping, type FieldType } from '../field-mappings';

export interface FieldResult {
  fieldName: string;
  selector: string;
  status: 'filled' | 'skipped' | 'failed';
  value?: string;
  error?: string;
}

const FILL_TIMEOUT_MS = 5000;

/**
 * Page Object Model for the G-28 target form.
 * Encapsulates all form interaction logic.
 */
export class FormA28Page {
  public filledFields: FieldResult[] = [];
  public skippedFields: FieldResult[] = [];
  public failedFields: FieldResult[] = [];

  constructor(
    private readonly page: Page,
    private readonly formUrl: string
  ) {}

  /**
   * Navigate to the target form.
   */
  async navigate(): Promise<void> {
    await this.page.goto(this.formUrl, { waitUntil: 'networkidle' });
  }

  /**
   * Fill the entire form with the provided data.
   */
  async fillForm(data: FormA28Data): Promise<void> {
    const mappings = getAllMappings();

    for (const mapping of mappings) {
      const value = this.getFieldValue(data, mapping.fieldName);
      await this.fillField(mapping, value);
    }
  }

  /**
   * Fill a single field based on its mapping and value.
   */
  private async fillField(mapping: FieldMapping, value: unknown): Promise<void> {
    const { fieldName, selector, fieldType } = mapping;

    // Skip if no value provided
    if (value === undefined || value === null || value === '') {
      this.skippedFields.push({
        fieldName,
        selector,
        status: 'skipped',
      });
      return;
    }

    try {
      await this.fillByType(fieldType, selector, value, fieldName);
      this.filledFields.push({
        fieldName,
        selector,
        status: 'filled',
        value: String(value),
      });
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : 'Unknown error';
      this.failedFields.push({
        fieldName,
        selector,
        status: 'failed',
        value: String(value),
        error: errorMessage,
      });
      // Log error but continue with other fields
      console.error(`Failed to fill field "${fieldName}": ${errorMessage}`);
    }
  }

  /**
   * Fill a field based on its type.
   */
  private async fillByType(
    fieldType: FieldType,
    selector: string,
    value: unknown,
    fieldName: string
  ): Promise<void> {
    switch (fieldType) {
      case 'text':
      case 'email':
      case 'tel':
        await this.fillTextField(selector, String(value));
        break;
      case 'select':
        await this.fillSelectField(selector, String(value));
        break;
      case 'checkbox':
        await this.fillCheckboxField(selector, Boolean(value));
        break;
      case 'radio':
        await this.fillRadioField(selector, String(value), fieldName);
        break;
      case 'date':
        await this.fillDateField(selector, String(value));
        break;
      default:
        throw new Error(`Unknown field type: ${fieldType}`);
    }
  }

  /**
   * Fill a text input field.
   */
  private async fillTextField(selector: string, value: string): Promise<void> {
    const element = await this.page.waitForSelector(selector, { timeout: FILL_TIMEOUT_MS });
    if (!element) {
      throw new Error(`Element not found: ${selector}`);
    }
    await element.fill(value);
  }

  /**
   * Fill a select dropdown field.
   */
  private async fillSelectField(selector: string, value: string): Promise<void> {
    const element = await this.page.waitForSelector(selector, { timeout: FILL_TIMEOUT_MS });
    if (!element) {
      throw new Error(`Element not found: ${selector}`);
    }
    await element.selectOption(value);
  }

  /**
   * Fill a checkbox field.
   */
  private async fillCheckboxField(selector: string, checked: boolean): Promise<void> {
    const element = await this.page.waitForSelector(selector, { timeout: FILL_TIMEOUT_MS });
    if (!element) {
      throw new Error(`Element not found: ${selector}`);
    }
    if (checked) {
      await element.check();
    } else {
      await element.uncheck();
    }
  }

  /**
   * Fill a radio button field.
   */
  private async fillRadioField(selector: string, value: string, fieldName: string): Promise<void> {
    // For radio buttons, we need to select the one with the matching value
    // Handle special cases like sex (M/F/X) and isSubjectToOrders (am/am_not)
    let radioSelector: string;

    if (fieldName === 'sex') {
      // Sex radio: value is M, F, or X
      radioSelector = `${selector}[value="${value}"]`;
    } else if (fieldName === 'isSubjectToOrders') {
      // Subject to orders: value is "am" or "am_not"
      radioSelector = `${selector}[value="${value}"]`;
    } else {
      radioSelector = `${selector}[value="${value}"]`;
    }

    const element = await this.page.waitForSelector(radioSelector, { timeout: FILL_TIMEOUT_MS });
    if (!element) {
      throw new Error(`Radio element not found: ${radioSelector}`);
    }
    await element.click();
  }

  /**
   * Fill a date input field.
   */
  private async fillDateField(selector: string, value: string): Promise<void> {
    const element = await this.page.waitForSelector(selector, { timeout: FILL_TIMEOUT_MS });
    if (!element) {
      throw new Error(`Element not found: ${selector}`);
    }
    // Date fields expect YYYY-MM-DD format
    await element.fill(value);
  }

  /**
   * Take a screenshot of the current page state.
   */
  async takeScreenshot(): Promise<Buffer> {
    return await this.page.screenshot({ fullPage: true });
  }

  /**
   * Verify that the form was not accidentally submitted.
   * Checks that we're still on the target form URL.
   */
  async verifyNoSubmit(): Promise<boolean> {
    const currentUrl = this.page.url();
    // Allow for URL variations (hash, query params) but ensure base URL matches
    return currentUrl.startsWith(this.formUrl.split('?')[0].split('#')[0]);
  }

  /**
   * Get a field value from the form data object.
   */
  private getFieldValue(data: FormA28Data, fieldName: string): unknown {
    // Use type assertion to access dynamic property
    return (data as Record<string, unknown>)[fieldName];
  }

  /**
   * Get a summary of the fill operation.
   */
  getSummary(): {
    total: number;
    filled: number;
    skipped: number;
    failed: number;
  } {
    return {
      total: this.filledFields.length + this.skippedFields.length + this.failedFields.length,
      filled: this.filledFields.length,
      skipped: this.skippedFields.length,
      failed: this.failedFields.length,
    };
  }
}

File: form-automation-service/app/page_objects/form_a28_page.py
"""Page Object Model for Form A-28 target form.

Target form: https://mendrika-alma.github.io/form-submission/

This module provides a clean abstraction over the form's DOM,
encapsulating all selectors and fill logic.
"""

import logging
from typing import Optional
from playwright.async_api import Page, TimeoutError as PlaywrightTimeout

from app.field_mapping import (
    FieldMapping,
    FieldType,
    ALL_FIELD_MAPPINGS,
    get_field_mapping,
)
from app.schemas import FieldResult, FieldStatus, FormA28Input

logger = logging.getLogger(__name__)


class FormA28Page:
    """Page Object for the A-28 immigration form."""

    def __init__(self, page: Page, form_url: str) -> None:
        self.page = page
        self.form_url = form_url
        self.filled_fields: list[FieldResult] = []
        self.skipped_fields: list[FieldResult] = []
        self.failed_fields: list[FieldResult] = []

    async def navigate(self) -> None:
        """Navigate to the form URL."""
        logger.info(f"Navigating to {self.form_url}")
        await self.page.goto(self.form_url, wait_until="networkidle")
        logger.info("Page loaded successfully")

    async def fill_form(self, data: FormA28Input) -> None:
        """Fill all form fields with provided data."""
        data_dict = data.model_dump(by_alias=False)

        for mapping in ALL_FIELD_MAPPINGS:
            value = data_dict.get(mapping.field_name)
            await self._fill_field(mapping, value)

    async def _fill_field(
        self, mapping: FieldMapping, value: Optional[str | bool]
    ) -> None:
        """Fill a single form field."""
        field_name = mapping.field_name
        selector = mapping.selector

        # Skip empty optional fields
        if value is None or value == "":
            if not mapping.required:
                self.skipped_fields.append(
                    FieldResult(
                        fieldName=field_name,
                        status=FieldStatus.SKIPPED,
                        value=None,
                        error="Empty optional field",
                    )
                )
                return
            else:
                self.failed_fields.append(
                    FieldResult(
                        fieldName=field_name,
                        status=FieldStatus.FAILED,
                        value=None,
                        error="Required field is empty",
                    )
                )
                return

        try:
            # Check if element exists
            element = await self.page.query_selector(selector)
            if not element:
                # Try alternative selectors for common patterns
                element = await self._try_alternative_selectors(mapping)
                if not element:
                    logger.warning(f"Element not found: {selector}")
                    self.skipped_fields.append(
                        FieldResult(
                            fieldName=field_name,
                            status=FieldStatus.SKIPPED,
                            value=str(value),
                            error=f"Selector not found: {selector}",
                        )
                    )
                    return

            # Fill based on field type
            if mapping.field_type == FieldType.TEXT:
                await self._fill_text_field(selector, str(value))
            elif mapping.field_type == FieldType.EMAIL:
                await self._fill_text_field(selector, str(value))
            elif mapping.field_type == FieldType.TEL:
                await self._fill_text_field(selector, str(value))
            elif mapping.field_type == FieldType.DATE:
                await self._fill_date_field(selector, str(value))
            elif mapping.field_type == FieldType.SELECT:
                await self._fill_select_field(selector, str(value))
            elif mapping.field_type == FieldType.CHECKBOX:
                await self._fill_checkbox_field(selector, bool(value))
            elif mapping.field_type == FieldType.RADIO:
                await self._fill_radio_field(selector, str(value))

            self.filled_fields.append(
                FieldResult(
                    fieldName=field_name,
                    status=FieldStatus.FILLED,
                    value=str(value),
                    error=None,
                )
            )
            logger.debug(f"Filled {field_name} with value: {value}")

        except PlaywrightTimeout:
            self.failed_fields.append(
                FieldResult(
                    fieldName=field_name,
                    status=FieldStatus.FAILED,
                    value=str(value),
                    error="Timeout waiting for element",
                )
            )
            logger.error(f"Timeout filling {field_name}")
        except Exception as e:
            self.failed_fields.append(
                FieldResult(
                    fieldName=field_name,
                    status=FieldStatus.FAILED,
                    value=str(value),
                    error=str(e),
                )
            )
            logger.error(f"Error filling {field_name}: {e}")

    async def _try_alternative_selectors(
        self, mapping: FieldMapping
    ) -> Optional[object]:
        """Try alternative selector patterns."""
        # Common alternative patterns based on field name
        field_name = mapping.field_name
        alternatives = [
            f'[name="{field_name}"]',
            f'[data-field="{field_name}"]',
            f'#{field_name.replace("_", "-")}',
            f'input[placeholder*="{field_name.replace("_", " ")}"]',
        ]

        for alt_selector in alternatives:
            element = await self.page.query_selector(alt_selector)
            if element:
                logger.info(f"Found alternative selector: {alt_selector}")
                return element
        return None

    async def _fill_text_field(self, selector: str, value: str) -> None:
        """Fill a text input field."""
        await self.page.fill(selector, value)

    async def _fill_date_field(self, selector: str, value: str) -> None:
        """Fill a date input field.

        Expects value in YYYY-MM-DD format.
        """
        await self.page.fill(selector, value)

    async def _fill_select_field(self, selector: str, value: str) -> None:
        """Fill a select dropdown."""
        await self.page.select_option(selector, value)

    async def _fill_checkbox_field(self, selector: str, checked: bool) -> None:
        """Set checkbox state."""
        if checked:
            await self.page.check(selector)
        else:
            await self.page.uncheck(selector)

    async def _fill_radio_field(self, selector: str, value: str) -> None:
        """Select a radio button by value."""
        # For radio buttons, we need to select the specific value
        radio_selector = f'{selector}[value="{value}"]'
        element = await self.page.query_selector(radio_selector)
        if element:
            await self.page.check(radio_selector)
        else:
            # Try clicking by label or other patterns
            await self.page.click(f'{selector}[value="{value}"]')

    async def take_screenshot(self) -> bytes:
        """Take a screenshot of the current page state."""
        return await self.page.screenshot(full_page=True)

    async def verify_no_submit(self) -> bool:
        """Verify that the form was NOT submitted.

        Check that we're still on the form page.
        """
        current_url = self.page.url
        return self.form_url in current_url or current_url == self.form_url
